crates/iceberg/src/transform/temporal.rs

```diff
@@ -0,0 +1,669 @@
+use super::TransformFunction;
+use crate::{Error, Result};
+use arrow::array::{
+    Array, Date64Array, TimestampMicrosecondArray, TimestampMillisecondArray,
+    TimestampNanosecondArray, TimestampSecondArray,
+};
+use arrow::compute::binary;
+use arrow::datatypes;
+use arrow::datatypes::DataType;
+use arrow::{
+    array::{ArrayRef, Date32Array, Int32Array},
+    compute::{month_dyn, year_dyn},
+};
+use chrono::Datelike;
+use std::sync::Arc;
+
+/// 719163 is the number of days from 0000-01-01 to 1970-01-01
+const EPOCH_DAY_FROM_CE: i32 = 719163;
+const DAY_PER_SECOND: f64 = 0.0000115741;
+const HOUR_PER_SECOND: f64 = 1_f64 / 3600.0;
+
+/// Extract a date or timestamp year, as years from 1970
+pub struct Year;
+
+impl TransformFunction for Year {
+    fn transform(&self, input: ArrayRef) -> Result<ArrayRef> {
+        let array = year_dyn(&input).map_err(|err| {
+            Error::new(
+                crate::ErrorKind::ArrowCompute,
```

Comment:
`ArrowCompute` is not an error code that can take actions.

The acutal error here is `{name} does not support`: https://arrow.apache.org/rust/src/arrow_arith/temporal.rs.html#452. But arrow will wrap all compute time error into `ArrowError::Compute`, so there is no good way for us to extract them. 

How about using `Unexpected` here? There is no action for our users to take.

Reply:
Sounds reasonable for me.

Reply:
Transform function is a internal function. For unsupport type, we should check it before call transform function and return error. We can panic directly if meet unsupport type in transform function. So can we replace it with `unreachable!` here.🤔

Reply:
Using `unreachable!()` at place where users should never reached is a good idea to me.